build-tools/packages/build-cli/src/commands/generate/typetests.ts

```diff
@@ -0,0 +1,377 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import path from "node:path";
+import {
+	type BrokenCompatTypes,
+	type Logger,
+	type Package,
+	type PackageJson,
+	type TestCaseTypeData,
+	type TypeData,
+	buildTestCase,
+	getFullTypeName,
+	getNodeTypeData,
+	typeOnly,
+} from "@fluidframework/build-tools";
+import { Flags } from "@oclif/core";
+import { PackageName } from "@rushstack/node-core-library";
+import * as changeCase from "change-case";
+import { mkdirSync, readJson, rmSync, writeFileSync } from "fs-extra";
+import * as resolve from "resolve.exports";
+import { ModuleKind, ModuleResolutionKind, Project, type SourceFile } from "ts-morph";
+import { PackageCommand } from "../../BasePackageCommand";
+import { ApiLevel, ensureDevDependencyExists, knownApiLevels } from "../../library";
+import { unscopedPackageNameString } from "./entrypoints";
+
+export default class GenerateTypetestsCommand extends PackageCommand<
+	typeof GenerateTypetestsCommand
+> {
+	static readonly description = "Generates type tests for a package or group of packages.";
+
+	static readonly flags = {
+		level: Flags.string({
+			description: "What API level to generate tests for.",
+			default: ApiLevel.internal,
+			options: knownApiLevels,
+		}),
+		outDir: Flags.directory({
+			description: "Where to emit the type tests file.",
+			default: "./src/test/types",
+		}),
+		outFile: Flags.string({
+			description: `File name for the generated type tests. The pattern '${unscopedPackageNameString}' within the value will be replaced with the unscoped name of this package in PascalCase.`,
+			default: `validate${unscopedPackageNameString}Previous.generated.ts`,
+		}),
+		publicFallback: Flags.boolean({
+			description:
+				"Use the public entrypoint as a fallback if the API at the requested level is not found.",
+			default: false,
+		}),
+		...PackageCommand.flags,
+	} as const;
+
+	protected async processPackage(pkg: Package): Promise<void> {
+		const { outDir, outFile } = this.flags;
+
+		// This cast is safe because oclif has already ensured only known ApiLevel values get to this point.
+		const level = this.flags.level as ApiLevel;
```

Comment:
I didn't do that because I got a typing error and didn't have time to chase it down:

```
Type 'typeof GenerateTypetestsCommand' does not satisfy the constraint 'typeof Command & { flags: { private: BooleanFlag<boolean>; scope: OptionFlag<string[] | undefined, CustomOptions>; skipScope: OptionFlag<string[] | undefined, CustomOptions>; ... 5 more ...; concurrency: OptionFlag<...>; }; }'.
  Type 'typeof GenerateTypetestsCommand' is not assignable to type 'typeof Command'.
    Types of property 'flags' are incompatible.
      Type '{ readonly private: BooleanFlag<boolean>; readonly scope: OptionFlag<string[] | undefined, CustomOptions>; readonly skipScope: OptionFlag<string[] | undefined, CustomOptions>; ... 9 more ...; readonly publicFallback: BooleanFlag<...>; }' is not assignable to type 'FlagInput<{ [flag: string]: any; }>'.
        Property 'level' is incompatible with index signature.
          Type 'FlagDefinition<ApiLevel, CustomOptions, { multiple: false; requiredOrDefaulted: true; }>' is not assignable to type 'Flag<any>'.ts(2344)
```

But looking at it now, my usage was wrong. It needs to be `Flags.custom<ApiLevel>(...options)()`. That missing `()` makes all the difference. I'll PR it.